File: packages/agent-client/typescript/src/protocol.ts
import { decode, encode } from "cbor-x";
import {
  ClientError, InboundFrame, MAX_FRAME_SIZE, readExactly, readRawFrame, validateIds,
  type ByteTransport, type EnvelopeCodec, type EstablishContext, type Established,
  type Protocol, type RawFrame, type SendMetadata,
} from "@microsandbox/protocol-client";
import { PROTOCOL_VERSION, messageFlags, supports } from "./message.js";

/** Prologue-selected wire form, separate from operation-generation gates. */
export type AgentWireFormat = "current" | "legacy_v1";
/** Ready metadata keeps optional historical fields and full-width timestamps. */
export type ReadyPayload = {
  boot_time_ns?: number | bigint;
  init_time_ns?: number | bigint;
  ready_time_ns?: number | bigint;
  agent_version?: string;
  [field: string]: unknown;
};

/** Metadata captured once, retaining the original ready frame and unknown fields. */
export class AgentReady {
  readonly agent: ReadyPayload;
  readonly negotiatedVersion: number;
  constructor(readonly wireFormat: AgentWireFormat, readonly frame: InboundFrame) {
    this.agent = frame.decodePayload<ReadyPayload>();
    this.negotiatedVersion = Math.min(wireFormat === "legacy_v1" ? 1 : PROTOCOL_VERSION, frame.protocolVersion);
  }
  get agentVersion(): string { return typeof this.agent?.agent_version === "string" ? this.agent.agent_version : ""; }
  get readyBytes(): Uint8Array { return this.frame.raw.body; }
  supports(wireName: string): boolean { return supports(wireName, this.negotiatedVersion); }
}

/** Preserve the existing agent encoder, including its map and byte-tag choices. */
export class AgentEnvelopeCodec implements EnvelopeCodec {
  encodePayload(value: unknown): Uint8Array { return encode(value); }
  encode(generation: number, wireName: string, payload: Uint8Array): Uint8Array {
    // Copy the finished envelope before cbor-x reuses its output allocation.
    return Uint8Array.from(encode({ v: generation, t: wireName, p: payload }));
  }
  decode(frame: RawFrame): InboundFrame {
    try {
      const value = decode(frame.body) as { v?: unknown; t?: unknown; p?: unknown } | null;
      // Agent decoders historically accept cbor-x's tagged Uint8Array payloads.
      // The stricter control codec is independent and does not inherit that rule.
      if (!value || !Number.isInteger(value.v) || Number(value.v) < 0 || Number(value.v) > 255
          || typeof value.t !== "string" || !(value.p instanceof Uint8Array)) throw new Error();
      return new InboundFrame(frame.id, frame.flags, Number(value.v), value.t, value.p, frame);
    } catch { throw new ClientError("invalid_data", "unknown"); }
  }
}

/** Relay handshake and agent gates; correlation and transport ownership are generic. */
export class AgentProtocol implements Protocol<AgentReady> {
  readonly reuseIds = false;
  async establish(transport: ByteTransport, context: EstablishContext): Promise<Established<AgentReady>> {
    const prologue = await readExactly(transport, 8, context.signal);
    const view = new DataView(prologue.buffer, prologue.byteOffset, prologue.byteLength);
    const first = view.getUint32(0), second = view.getUint32(4);
    // This is the same disambiguation as the existing Rust legacy relay adapter.
    const legacy = second >= 5 && second <= MAX_FRAME_SIZE && (first === 0 || first >= second);
    const ids = legacy
      ? { start: Math.min(first + 1, 0xffffffff), endExclusive: Math.min(first + Math.floor(0xffffffff / 16), 0xffffffff) }
      : { start: Math.max(first, 1), endExclusive: second };
    validateIds(ids);
    let raw: RawFrame;
    if (legacy) {
      const header = await readExactly(transport, 5, context.signal);
      raw = {
        id: new DataView(header.buffer).getUint32(0), flags: header[4]!,
        body: await readExactly(transport, second - 5, context.signal),
      };
    } else {
      const queued = await readRawFrame(transport);
      if (!queued) throw new ClientError("peer_closed");
      raw = queued.frame;
    }
    const codec = new AgentEnvelopeCodec();
    const frame = codec.decode(raw);
    if (frame.type !== "core.ready") throw new ClientError("invalid_data");
    return { transport, codec, ids, ready: new AgentReady(legacy ? "legacy_v1" : "current", frame), limits: context.limits };
  }
  prepare(ready: AgentReady, wireName: string): SendMetadata {
    if (!ready.supports(wireName)) throw new ClientError("unsupported_operation");
    return { generation: ready.wireFormat === "legacy_v1" ? 1 : PROTOCOL_VERSION, flags: messageFlags(wireName) };
  }
}

File: packages/agent-client/rust/lib/protocol.rs
//! Agent relay setup and metadata over the shared framed router.

use std::sync::Arc;

use microsandbox_protocol::{
    codec::{self, MAX_FRAME_SIZE, RawFrame},
    core::Ready,
    message::{FRAME_HEADER_SIZE, MessageType, PROTOCOL_VERSION},
};
use microsandbox_protocol_client::{
    BoxFuture, BoxTransport, CborEnvelopeCodec, ClientError, ClientResult, ConnectOptions,
    ErrorKind, Established, IdRange, Protocol, SendMetadata,
};
use tokio::io::AsyncReadExt;

//--------------------------------------------------------------------------------------------------
// Constants
//--------------------------------------------------------------------------------------------------

/// Protocol generation used by the supported pre-0.5 agent wire path.
pub const LEGACY_PROTOCOL_VERSION: u8 = 1;
const LEGACY_RELAY_ID_RANGE_STEP: u32 = u32::MAX / 16;

//--------------------------------------------------------------------------------------------------
// Types
//--------------------------------------------------------------------------------------------------

/// Agent relay protocol specialization; it owns no router or SDK services.
pub struct AgentProtocol;

/// Wire form selected from the relay prologue, separate from capability gates.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AgentWireFormat {
    /// Current relay prologue and envelope generation.
    Current,
    /// Supported pre-0.5 relay prologue and generation-one envelope.
    LegacyV1,
}

/// Immutable metadata captured once during relay setup.
#[derive(Debug, Clone)]
pub struct AgentReady {
    /// Relay wire form selected during the prologue.
    pub wire_format: AgentWireFormat,
    /// Smaller of host and peer generations, used for known-operation gates.
    pub negotiated_version: u8,
    /// Decoded original `core.ready` payload.
    pub agent: Ready,
    ready_body: Vec<u8>,
}

//--------------------------------------------------------------------------------------------------
// Methods
//--------------------------------------------------------------------------------------------------

impl AgentWireFormat {
    /// Envelope generation emitted on this wire form.
    ///
    /// Current peers historically receive the host generation even when the
    /// feature gate negotiated lower; preserve that existing byte contract.
    pub fn version(self) -> u8 {
        match self {
            Self::Current => PROTOCOL_VERSION,
            Self::LegacyV1 => LEGACY_PROTOCOL_VERSION,
        }
    }
}

impl AgentReady {
    /// Exact original ready envelope, including unknown fields.
    pub fn ready_bytes(&self) -> &[u8] {
        &self.ready_body
    }

    /// Whether the connected generation supports this known message.
    pub fn supports(&self, message_type: MessageType) -> bool {
        message_type.is_available_at(self.negotiated_version)
    }

    /// Self-reported package version, empty on older agents lacking the field.
    pub fn agent_version(&self) -> &str {
        &self.agent.agent_version
    }

    /// Whether the connection selected the supported pre-0.5 wire form.
    pub fn is_legacy_protocol(&self) -> bool {
        self.wire_format == AgentWireFormat::LegacyV1
    }
}

impl AgentProtocol {
    /// Validate a known operation against a separately retained generation.
    pub fn ensure_version_compat_for(
        message_type: MessageType,
        negotiated: u8,
    ) -> ClientResult<()> {
        if message_type.is_available_at(negotiated) {
            Ok(())
        } else {
            Err(ClientError::new(ErrorKind::UnsupportedOperation))
        }
    }
}

//--------------------------------------------------------------------------------------------------
// Trait Implementations
//--------------------------------------------------------------------------------------------------

impl Protocol for AgentProtocol {
    // The generation-eight relay permanently retires completed correlations.
    const REUSE_IDS: bool = false;
    type Ready = AgentReady;

    fn establish(
        mut stream: BoxTransport,
        options: ConnectOptions,
    ) -> BoxFuture<'static, ClientResult<Established<AgentReady>>> {
        Box::pin(async move {
            // The engine bounds this entire future with one setup deadline.
            // Eight bytes disambiguate current [min,max] and legacy [offset,len].
            let mut prologue = [0u8; 8];
            stream.read_exact(&mut prologue).await?;
            let first = u32::from_be_bytes(prologue[..4].try_into().unwrap());
            let second = u32::from_be_bytes(prologue[4..].try_into().unwrap());
            let legacy = (FRAME_HEADER_SIZE as u32..=MAX_FRAME_SIZE).contains(&second)
                && (first == 0 || first >= second);
            let (ids, frame, wire_format) = if legacy {
                let frame = read_after_prefix(&mut stream, second).await?;
                (
                    IdRange {
                        start: first.saturating_add(1),
                        end_exclusive: first.saturating_add(LEGACY_RELAY_ID_RANGE_STEP).into(),
                    },
                    frame,
                    AgentWireFormat::LegacyV1,
                )
            } else {
                // Current ranges can start at zero, but ID zero is setup-only.
                let ids = IdRange {
                    start: first.max(1),
                    end_exclusive: second.into(),
                };
                ids.validate()?;
                let frame = codec::read_raw_frame(&mut stream)
                    .await
                    .map_err(|_| ClientError::new(ErrorKind::InvalidData))?;
                (ids, frame, AgentWireFormat::Current)
            };
            ids.validate()?;
            // Keep the historical ready decoder and feature-generation rule.
            let ready_message = codec::raw_frame_to_message(frame.clone())
                .map_err(|_| ClientError::new(ErrorKind::InvalidData))?;
            if ready_message.t != MessageType::Ready {
                return Err(ClientError::new(ErrorKind::InvalidData));
            }
            let agent = ready_message
                .payload::<Ready>()
                .map_err(|_| ClientError::new(ErrorKind::InvalidData))?;
            let ready = AgentReady {
                wire_format,
                negotiated_version: wire_format.version().min(ready_message.v),
                agent,
                ready_body: frame.body,
            };
            if ready.is_legacy_protocol() {
                tracing::warn!(
                    "agent client: legacy pre-0.5 exec protocol; filesystem operations require a newer agent"
                );
            }
            Ok(Established {
                transport: stream,
                codec: Arc::new(CborEnvelopeCodec),
                ids,
                ready,
                limits: options.limits,
            })
        })
    }

    fn prepare(ready: &AgentReady, wire_name: &str) -> ClientResult<SendMetadata> {
        let flags = match MessageType::from_wire_str(wire_name) {
            Some(message_type) => {
                Self::ensure_version_compat_for(message_type, ready.negotiated_version)?;
                message_type.flags()
            }
            // Dynamic names need no schema registration. Raw sends also support
            // caller-selected flags and generations.
            None => 0,
        };
        Ok(SendMetadata {
            generation: ready.wire_format.version(),
            flags,
        })
    }
}

//--------------------------------------------------------------------------------------------------
// Functions
//--------------------------------------------------------------------------------------------------

async fn read_after_prefix(stream: &mut BoxTransport, length: u32) -> ClientResult<RawFrame> {
    if !(FRAME_HEADER_SIZE as u32..=MAX_FRAME_SIZE).contains(&length) {
        return Err(ClientError::new(ErrorKind::InvalidData));
    }
    let mut header = [0u8; FRAME_HEADER_SIZE];
    stream.read_exact(&mut header).await?;
    let mut body = vec![0; length as usize - FRAME_HEADER_SIZE];
    stream.read_exact(&mut body).await?;
    Ok(RawFrame {
        id: u32::from_be_bytes(header[..4].try_into().unwrap()),
        flags: header[4],
        body,
    })
}
